### PR TITLE
feat(datasource/rubygems): Fetch metadata for rubygems

### LINKED ISSUE #3670: Add back metadata for rubygems.org packages to get release notes / changelog
**What would you like Renovate to be able to do?**

Add back rubygems package and version metadata (e.g. package home page, version release date, and ruby compatibility) that was "lost" in #3669.

**Describe the solution you'd like**

Rely on the CDN to get a full view of every version of every package, but when we detect it's changed then we should do an API-based fetch of the package and versions.

I think this should work well:

1. Query the versions file to get a full list of releases for the package
2. Query our cache for full metadata of the package. If the cache is empty for this package, or the versions list does not match what we found in the CDN file, then we refresh and overwrite it. No expiry date, but we should add a "last fetched" field of our own.

Also:
- We need to make sure we don't hit the Rubygems API [rate limits](https://guides.rubygems.org/rubygems-org-api/#rate-limits). Rate limiting per second needs to be built into the fetcher, something conservative like 5 per second.
- We need to defensively program against querying the API every time if there's a permanent mismatch. Our internal "last fetched" date can help wtih this

**Describe alternatives you've considered**

The `/info/*` endpoint for packages is CDN-based but is missing the release date of each version, which I think is really important. Therefore we need to query the API endpoints for the full data.

Another consideration was whether we need to re-query the package metadata (source URL, homepage, etc) periodically but I think it makes sense to only refresh it if there's a new *version* published, and this approach will be much more efficient.

**Additional context**

See #3373

### PR DESCRIPTION
**Pull request by Tranquility**
## Changes

Use rubygems datasource that fetches metadata and add `maxRequestsPerSecond` to not exceed api rate limit. I don't know if this is the right approach so I did not delete the versions-datasource yet.

## Context

closes #3670

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
this makes performance much more worse. I'm not sure we can accept this change. 🤔

**Comment by Tranquility** [comment]
> this makes performance much more worse. I'm not sure we can accept this change. 🤔

Yes, that's why I wasn't sure about the approach. What would be the solution? Isn't this only a problem when it runs for the first time because it is cached afterwards? 
How could we only use the datasource with the metadata when we really need to?

**Comment by Tranquility** [comment]
@rarkins do you have any suggestions how to improve this?

File: lib/config/presets/internal/workarounds.ts [hunk 1]
```diff
@@ -32,6 +33,16 @@ export const presets: Record<string, Preset> = {
...
+  doNotExceedRubygemsRateLimit: {
+    description:
+      'Limit concurrent requests to not exceed the rubygems rate limit.',
+    hostRules: [
+      {
+        matchHost: 'rubygems.org',
+        maxRequestsPerSecond: 15,
+      },
+    ],
+  },
```
Comment (rarkins) [inline lib/config/presets/internal/workarounds.ts:45]: I think we should have this type of thing as default/embedded when we know a certain registry doesn't want to be overloaded

**Review by rarkins — CHANGES_REQUESTED** [review]
I think what we should be doing is _supplementing_ the versions datasource with metadata. The metadata for any package (e.g. source URL, home page) can be cached a long time, such as 24 hours or even a week. Here is some prior art from Rust crates.io: https://github.com/renovatebot/renovate/blob/a91ca62b466b57aa9a18f4e15bd04eb5552278b1/lib/modules/datasource/crate/index.ts#L116-L127

**Comment by rarkins** [comment]
Will be replaced by some work by @zharinov

**Pull request closed without merging**